### PR TITLE
Scheduled Reports | X o'clock UTC >> XX:00 UTC

### LINKED ISSUE #21118: X o'clock UTC >> XX:00 UTC
## Summary

"X o'clock" is very informal and not very clear, since it's not a common pattern to show time. I'd suggest changing it to the more common pattern "XX:00 UTC" (of course including any time conversions that we use)
image

## Your Environment

* Matomo Version: Matomo Cloud (4.15)

### PR DESCRIPTION
**Pull request by chippison**
### Description
DEV-17057

We want to change the labeling in our Scheduled report for when to send the report. 

fixes #21118

### Checklist

- [NA] I have understood, reviewed, and tested all AI outputs before use
- [NA] All AI instructions respect security, IP, and privacy rules

### Review

- [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
- [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behaviour of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
- [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
- [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
- [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
- [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
- [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
- [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
- [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
- [ ] [New documentation added or existing documentation updated](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)

### CODE REVIEW

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue [hunk 1]
```diff
@@ -617,15 +617,16 @@ export default defineComponent({
     reportHours() {
       const hours: Option[] = [];
       for (let i = 0; i < 24; i += 1) {
+        const paddedHour = i < 10 ? `0${i}` : `${i}`;
```
Comment (tzi) [inline plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: There is a [String.padStart()](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Global_Objects/String/padStart) method for this

```suggestion
        const paddedHour = `${i}`.padStart(2, '0');
```

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue [hunk 2]
```diff
@@ -636,7 +637,13 @@ export default defineComponent({
         this.report.hour as string,
         -this.timeZoneDifferenceInHours,
       );
-      return translate('ScheduledReports_ReportHourWithUTC', [reportHour]);
+      const reportHourFloat = parseFloat(reportHour);
+      const hours = Math.floor(reportHourFloat);
+      const paddedHour = hours < 10 ? `0${hours}` : `${hours}`;
```
Comment (tzi) [inline plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: Same here

```suggestion
      const paddedHour = `${hours}`.padStart(2, '0');
```

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue [hunk 3]
```diff
@@ -636,7 +637,13 @@ export default defineComponent({
...
+      const reportHourUtcLabel = translate('ScheduledReports_ReportHourWithUTC', [`${paddedHour}:00`]);
```
Comment (tzi) [inline plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: I don’t know when to use an array for the translation parameters and when we can use several parameters instead. I’ve seen both in the code base. Do you have any hint?

File: tests/UI/specs/UIIntegration_spec.js [hunk 4]
```diff
@@ -729,8 +729,23 @@ describe("UIIntegrationTest", function () { // TODO: Rename to Piwik?
...
         });
+        it('should load scheduled reports showing report hour description when timezone is not utc', async function () {
+          await testEnvironment.callApi('SitesManager.updateSite', {
+            idSite: 1,
+            timezone: 'Asia/Kolkata', // UTC+05:30
+          });
+          await page.goto("?" + generalParams + "&module=ScheduledReports&action=index");
+          await page.click('.entityTable tr:nth-child(11) button[title="Edit"]');
+
+          expect(await screenshotPageWrap()).to.matchImage('email_reports_editor_tz');
```
Comment (tzi) [inline tests/UI/specs/UIIntegration_spec.js]: It could be great to only check how the time is displayed instead of creating a new screenshot here.

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue (hunk 3 above)
Comment (sgiehl) [reply plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: FYI: Using `${paddedHour}:00` is actually ignoring i18n. It might be an acceptable trade of for now, but actually we would need to
* Adjust the Intl importer, to also import the time format `HH:mm` for each language
* Adjust our formatters to handle that (php & js)
* Use the formatter here

Comment (tzi) [reply plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: I’m not sure to follow what you want to say by "FYI: Using ${paddedHour}:00 is actually ignoring i18n.". Can you give more explanations? 🙏

Comment (sgiehl) [reply plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: Not every language uses the notation `HH:mm`, some might only use `H:mm`. Furthermore a lot languages are not using `:`, but would actually use a different char instead.

Comment (chippison) [reply plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: > I don’t know when to use an array for the translation parameters and when we can use several parameters instead. I’ve seen both in the code base. Do you have any hint?

I think the translation function will accept an array or string as parameter.

Comment (chippison) [reply plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: >FYI: Using ${paddedHour}:00 is actually ignoring i18n. It might be an acceptable trade of for now, but actually we would need to
>Adjust the Intl importer, to also import the time format HH:mm for each language
>Adjust our formatters to handle that (php & js)
>Use the formatter here

@sgiehl 
Hi Steffan, 
since the issue here was just for a text change and no one actually asked/complained about i18n for this ticket, is it all good to do this on a separate issue and just prioritize it at a different time?

File: tests/UI/specs/UIIntegration_spec.js (hunk 4 above)
Comment (chippison) [reply tests/UI/specs/UIIntegration_spec.js]: @tzi 
The test here was actually to see if the help text/description shows up or not when timezone is not in UTC. 
I'll have a go at targeting just the help text and checking if its showing up or not

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue (hunk 1 above)
Comment (chippison) [reply plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: Thanks for the padStart suggestion. Adding this in

File: tests/UI/specs/UIIntegration_spec.js (hunk 4 above)
Comment (chippison) [reply tests/UI/specs/UIIntegration_spec.js]: Done

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue (hunk 3 above)
Comment (sgiehl) [reply plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: Sure. The previous format didn't have any part that could have been formatted differently in terms of i18n. It was just a simple number. The new format however could at some point bring up 18n issues around this.
We don't need to change that now, I just wanted to create some awareness around it.

File: plugins/ScheduledReports/lang/en.json [hunk 5]
```diff
@@ -25,8 +25,8 @@
         "PleaseFindBelow": "Please find below your %1$s report for %2$s.",
         "PluginDescription": "Create custom reports and schedule them to be emailed daily, weekly or monthly to one or several people. Several report formats are supported (html, pdf, csv, images).",
         "ReportFormat": "Report Format",
-        "ReportHour": "Send report at %s o'clock",
-        "ReportHourWithUTC": "%s o'clock UTC",
+        "ReportHour": "Send report at %s (local time)",
+        "ReportHourWithUTC": "%s UTC",
```
Comment (sgiehl) [inline plugins/ScheduledReports/lang/en.json]: Even though the sense of those translations didn't change, we should consider changing the key nevertheless.
Otherwise other languages might end up displaying a mixture (e.g. `13:00 o'clock UTC`)

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue (hunk 3 above)
Comment (chippison) [reply plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: Thanks for this info Steffan

File: plugins/ScheduledReports/lang/en.json (hunk 5 above)
Comment (chippison) [reply plugins/ScheduledReports/lang/en.json]: Changed the keys for this now

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue [hunk 6]
```diff
@@ -109,7 +109,7 @@
           name="report_hour"
           :model-value="report.hour"
           @update:model-value="$emit('change', { prop: 'hour', value: $event })"
-          :title="translate('ScheduledReports_ReportHour', 'X')"
+          :title="translate('ScheduledReports_ReportHourLocal', 'X')"
```
Comment (sgiehl) [inline plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: It looks like this translation key is only used here. Why should we use the placeholder, if the value is static? Also `Send report at X (local time)` doesn't look nice. Should we maybe just use `Send report at (local time)`?

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue [hunk 7]
```diff
@@ -636,7 +637,13 @@ export default defineComponent({
         this.report.hour as string,
         -this.timeZoneDifferenceInHours,
       );
-      return translate('ScheduledReports_ReportHourWithUTC', [reportHour]);
+      const reportHourFloat = parseFloat(reportHour);
+      const hours = Math.floor(reportHourFloat);
+      const paddedHour = `${hours}`.padStart(2, '0');
+      const reportHourUtcLabel = translate('ScheduledReports_ReportHourWithUtcOnly', [`${paddedHour}:00`]);
```
Comment (sgiehl) [inline plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: Incorrect UTC time formatting for half-hour offsets. reportHourUtc floors the hour and always formats :00, so 10.5 becomes 10:00 instead of 10:30. This is wrong for time zones like Asia/Kolkata.

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue [hunk 8]
```diff
@@ -636,7 +637,13 @@ export default defineComponent({
...
+      return `${translate('ScheduledReports_ReportWillBeSentAt')}
+      ${translate('ScheduledReports_ReportHourEqualsUtc', [reportHourUtcLabel])}
+      ${translate('ScheduledReports_NoteDeliveryTime')}`;
```
Comment (sgiehl) [inline plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: The help text is composed as a single string with embedded newlines; depending on rendering, it may collapse into odd spacing. Not breaking, but might affect readability in some browsers.

File: tests/UI/specs/UIIntegration_spec.js [hunk 9]
```diff
@@ -729,6 +729,26 @@ describe("UIIntegrationTest", function () { // TODO: Rename to Piwik?
...
+          // put back default timezone
+          await testEnvironment.callApi('SitesManager.updateSite', {
+            idSite: 1,
+            timezone: 'UTC',
+          });
+          await page.goto("?" + generalParams + "&module=ScheduledReports&action=index");
+          await page.click('.entityTable tr:nth-child(11) button[title="Edit"]');
+          await page.waitForNetworkIdle();
+          expect(await page.evaluate(() => $('#reportHourHelpText').is(':visible'))).to.be.false;
+        });
```
Comment (sgiehl) [inline tests/UI/specs/UIIntegration_spec.js:771]: UI test only checks visibility, not correctness of the UTC text. It won’t catch the half‑hour formatting bug.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Issues identified by Codex

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue (hunk 7 above)
Comment (chippison) [reply plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: Hi @sgiehl ,
I saw this when reviewing from codex as well, but when I checked with what was happening in `5.x-dev`, it was also not showing the minutes and just always the hours. 

Like:
```
Choose 7:30 -> 22 o'clock
```

We also have code in reportHours that changes our selectbox to add '.5' to our selections so that we also ensure we choose something with 'xx:30' when we are in timezone with 'UTC+30'
```
if ((this.timeZoneDifferenceInHours * 2) % 2 !== 0) {
  hours.push({
    key: `${i}.5`,
    value: `${paddedHour}:30`,
  });
} else {
  hours.push({
    key: `${i}`,
    value: `${paddedHour}:00`,
  });
}
```
Meaning, if we have a timezone with UTC+30, we are forced to select '00:30, 01:30, etc' instead of '00:00, 01:00'

I thought that because of these, this was what it was intending to do and we always 'try' to send the report at the hour and not in 30 mins increment. 

Caveat: 
This will not work for UTC+45 timezones though, but after checking its only Nepal and Chatham Islands that seem to use this timezone, but let me know if we want to make changes so that we cover this edge case

File: tests/UI/specs/UIIntegration_spec.js (hunk 9 above)
Comment (chippison) [reply tests/UI/specs/UIIntegration_spec.js:771]: Added test for actual time when the timezone value is not UTC. If it is UTC, we do not check the time value since we just hide the help text anyway

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue (hunk 7 above)
Comment (chippison) [reply plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: I've just added the fix so that UTC+45 timezones are covered

File: tests/UI/specs/UIIntegration_spec.js [hunk 10]
```diff
@@ -729,6 +729,46 @@ describe("UIIntegrationTest", function () { // TODO: Rename to Piwik?
...
+        it('should load scheduled reports showing report hour description for UTC+30 and UTC+45', async function () {
+          const timezoneCases = [
+            { timezone: 'Asia/Kolkata', value: 'string:10.5' }, // UTC+05:30
+            { timezone: 'Asia/Kathmandu', value: 'string:10.75' }, // UTC+05:45
+          ];
+
+          try {
+            for (const timezoneCase of timezoneCases) {
+              await testEnvironment.callApi('SitesManager.updateSite', {
+                idSite: 1,
```
Comment (sgiehl) [inline tests/UI/specs/UIIntegration_spec.js]: The UI test fixture used here contains around 10 different sites. Most of the tests are performed with site id 1, as it contains most of the tracking data.
Therefor it could be wise to choose a different site for manipulation here. I know it is reset in the finally part, but if that might fail for any weird reason it would currently have an effect on tests running afterwards.
To reduce the possible problems around that it might be better to use e.g. id 4, which currently seems not to be used anywhere else.
I'll just change that myself, so we can merge it.

**Review by sgiehl — DISMISSED** [review]

**Review by sgiehl — APPROVED** [review]
